## Introduction

In this project, data set of submissions to popular technology site Hacker News was worked on. Hacker News is a site started by the startup incubator Y Combinator, where user-submitted stories (known as "posts") are voted and commented upon, similar to reddit. Hacker News is extremely popular in technology and startup circles, and posts that make it to the top of Hacker News' listings can get hundreds of thousands of visitors as a result.

Below are descriptions of the columns:
* **id**: The unique identifier from Hacker News for the post
* **title:** The title of the post
* **url:** The URL that the posts links to, if it the post has a URL
* **num_points:** The number of points the post acquired, calculated as the total number of upvotes minus the total number of downvotes
* **num_comments:** The number of comments that were made on the post
* **author:** The username of the person who submitted the post
* **created_at:** The date and time at which the post was submitted

I am specifically interested in posts whose titles begin with either **Ask HN** or **Show HN**. Users submit Ask HN posts to ask the Hacker News community a specific question. Likewise, users submit **Show HN** posts to show the Hacker News community a project, product, or just generally something interesting.

## Aim and Objectives

I'll compare these two types of posts to determine the following:

* Do **Ask HN** or **Show HN** receive more comments on average?
* Do posts created at a certain time receive more comments on average?


In [15]:
from csv import reader

hn_open = open ('hacker_news.csv', encoding="utf8")
hn_read = reader(hn_open)
hn = list (hn_read)
print (hn[0:4])

[['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at'], ['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20']]


## Separating the Header from the List of List

In [16]:
header = hn[0] # header
print (header)

['id', 'title', 'url', 'num_points', 'num_comments', 'author', 'created_at']


In [17]:
hn = hn[1:]  # dataset body
print (hn[:4])

[['12224879', 'Interactive Dynamic Video', 'http://www.interactivedynamicvideo.com/', '386', '52', 'ne0phyte', '8/4/2016 11:52'], ['10975351', 'How to Use Open Source and Shut the Fuck Up at the Same Time', 'http://hueniverse.com/2016/01/26/how-to-use-open-source-and-shut-the-fuck-up-at-the-same-time/', '39', '10', 'josep2', '1/26/2016 19:30'], ['11964716', "Florida DJs May Face Felony for April Fools' Water Joke", 'http://www.thewire.com/entertainment/2013/04/florida-djs-april-fools-water-joke/63798/', '2', '1', 'vezycash', '6/23/2016 22:20'], ['11919867', 'Technology ventures: From Idea to Enterprise', 'https://www.amazon.com/Technology-Ventures-Enterprise-Thomas-Byers/dp/0073523429', '3', '1', 'hswarna', '6/17/2016 0:01']]


## Extracting Ask HN and Show HN Posts

Separation of posts beginning with Ask HN and Show HN (and case variations) into two different lists next.

In [18]:
ask_posts = []
show_posts = [] 
other_posts = []

for row in hn:
    title = row[1]
    title = title.lower()
    
    if title.startswith('ask hn'):
        ask_posts.append(row)
    elif title.startswith('show hn'):
        show_posts.append(row)
    else:
        other_posts.append(row)
        
num_ask_post = len(ask_posts)
num_show_post = len(show_posts)
num_other_post = len(other_posts)

print (num_ask_post)
print (num_show_post)
print (num_other_post)

1744
1162
17194


Below are the first five rows in the ask_posts list of lists:

In [19]:
print (ask_posts[:5])


[['12296411', 'Ask HN: How to improve my personal website?', '', '2', '6', 'ahmedbaracat', '8/16/2016 9:55'], ['10610020', 'Ask HN: Am I the only one outraged by Twitter shutting down share counts?', '', '28', '29', 'tkfx', '11/22/2015 13:43'], ['11610310', 'Ask HN: Aby recent changes to CSS that broke mobile?', '', '1', '1', 'polskibus', '5/2/2016 10:14'], ['12210105', 'Ask HN: Looking for Employee #3 How do I do it?', '', '1', '3', 'sph130', '8/2/2016 14:20'], ['10394168', 'Ask HN: Someone offered to buy my browser extension from me. What now?', '', '28', '17', 'roykolak', '10/15/2015 16:38']]


Below are the first five rows in the ask_posts list of lists:

In [37]:
print(show_posts[:5])

[['10627194', 'Show HN: Wio Link  ESP8266 Based Web of Things Hardware Development Platform', 'https://iot.seeed.cc', '26', '22', 'kfihihc', '11/25/2015 14:03'], ['10646440', 'Show HN: Something pointless I made', 'http://dn.ht/picklecat/', '747', '102', 'dhotson', '11/29/2015 22:46'], ['11590768', 'Show HN: Shanhu.io, a programming playground powered by e8vm', 'https://shanhu.io', '1', '1', 'h8liu', '4/28/2016 18:05'], ['12178806', 'Show HN: Webscope  Easy way for web developers to communicate with Clients', 'http://webscopeapp.com', '3', '3', 'fastbrick', '7/28/2016 7:11'], ['10872799', 'Show HN: GeoScreenshot  Easily test Geo-IP based web pages', 'https://www.geoscreenshot.com/', '1', '9', 'kpsychwave', '1/9/2016 20:45']]


## Calculating the Average Number of Comments for Ask HN and Show HN Posts

Determining which of ask posts or show posts receive more comments on average.

In [21]:
# total number of comments and avg. comments in ask posts

total_ask_comments = 0  

for row in ask_posts:
    ask_num_comments = row[4]
    ask_num_comments = int(ask_num_comments)
    total_ask_comments += ask_num_comments

avg_ask_comments = total_ask_comments/num_ask_post

print (total_ask_comments)
print (avg_ask_comments)


24483
14.038417431192661


In [22]:
# total number of comments and avg. comments in show posts

total_show_comments = 0  

for row in show_posts:
    show_num_comments = row[4]
    show_num_comments = int(show_num_comments)
    total_show_comments += show_num_comments
avg_show_comments = total_show_comments/num_show_post

print (total_show_comments)
print (avg_show_comments)

11988
10.31669535283993


On the average ask post receive more comments than show post. From the result obtained, average show comment is 10.31669535283993 while that of ask comment 14.038417431192661


## Finding the Amount of Ask Posts and Comments by Hour Created

In [35]:
# Calculate the amount of ask posts created during each hour of day and the number of comments received.
# counts_by_hour: contains the number of ask posts created during each hour of the day.
#comments_by_hour: contains the corresponding number of comments ask posts created at each hour received

import datetime as dt

result_list = []

for row in ask_posts:
    result_list.append([row[6],int(row[4])])
    

comments_by_hour = {}
counts_by_hour = {}
date_format = "%m/%d/%Y %H:%M"

for row_list in result_list:
    date = row_list[0]
    comment = row_list[1]
    time = dt.datetime.strptime(date,date_format).strftime("%H")
    
    if time in counts_by_hour:
        counts_by_hour[time] += 1
        comments_by_hour[time] += comment
    else:
        counts_by_hour[time] = 1
        comments_by_hour[time] = comment
        
comments_by_hour


{'09': 251,
 '13': 1253,
 '10': 793,
 '14': 1416,
 '16': 1814,
 '23': 543,
 '12': 687,
 '17': 1146,
 '15': 4477,
 '21': 1745,
 '20': 1722,
 '02': 1381,
 '18': 1439,
 '03': 421,
 '05': 464,
 '19': 1188,
 '01': 683,
 '22': 479,
 '08': 492,
 '04': 337,
 '00': 447,
 '06': 397,
 '07': 267,
 '11': 641}

## Calculating the Average Number of Comments for Ask HN Posts by Hour

In [39]:
# Calculate the average amount of comments `Ask HN` posts created at each hour of the day receive.

avg_by_hour = []
for hr in comments_by_hour:
    avg_by_hour.append([hr, comments_by_hour[hr]/counts_by_hour[hr] ])
    
avg_by_hour

[['09', 5.5777777777777775],
 ['13', 14.741176470588234],
 ['10', 13.440677966101696],
 ['14', 13.233644859813085],
 ['16', 16.796296296296298],
 ['23', 7.985294117647059],
 ['12', 9.41095890410959],
 ['17', 11.46],
 ['15', 38.5948275862069],
 ['21', 16.009174311926607],
 ['20', 21.525],
 ['02', 23.810344827586206],
 ['18', 13.20183486238532],
 ['03', 7.796296296296297],
 ['05', 10.08695652173913],
 ['19', 10.8],
 ['01', 11.383333333333333],
 ['22', 6.746478873239437],
 ['08', 10.25],
 ['04', 7.170212765957447],
 ['00', 8.127272727272727],
 ['06', 9.022727272727273],
 ['07', 7.852941176470588],
 ['11', 11.051724137931034]]


## Sorting and Printing Values from a List of Lists

In [44]:
swap_avg_by_hour = []

for row in avg_by_hour:
    swap_avg_by_hour.append([row[1], row[0]])

sorted_swap = sorted(swap_avg_by_hour, reverse = True )

sorted_swap

    
    

[[38.5948275862069, '15'],
 [23.810344827586206, '02'],
 [21.525, '20'],
 [16.796296296296298, '16'],
 [16.009174311926607, '21'],
 [14.741176470588234, '13'],
 [13.440677966101696, '10'],
 [13.233644859813085, '14'],
 [13.20183486238532, '18'],
 [11.46, '17'],
 [11.383333333333333, '01'],
 [11.051724137931034, '11'],
 [10.8, '19'],
 [10.25, '08'],
 [10.08695652173913, '05'],
 [9.41095890410959, '12'],
 [9.022727272727273, '06'],
 [8.127272727272727, '00'],
 [7.985294117647059, '23'],
 [7.852941176470588, '07'],
 [7.796296296296297, '03'],
 [7.170212765957447, '04'],
 [6.746478873239437, '22'],
 [5.5777777777777775, '09']]

In [50]:

# Sort the values and print the the 5 hours with the highest average comments.
print("Top 5 Hours for 'Ask HN' Comments")

for avg, hr in sorted_swap[:5]:
    print("{}:{:.2f}average comments per post".format(dt.datetime.strptime(hr,"%H").strftime("%H:%M"), avg) )
    
    

Top 5 Hours for 'Ask HN' Comments
15:00:38.59average comments per post
02:00:23.81average comments per post
20:00:21.52average comments per post
16:00:16.80average comments per post
21:00:16.01average comments per post


The hour that receives the most comments per post on average is 15:00, with an average of 38.59 comments per post. There's about a 60% increase in the number of comments between the hours with the highest and second highest average number of comments.

## Conclusion

In this project, I analyzed ask posts and show posts to determine which type of post and time receive the most comments on average. Based on my analysis, to maximize the amount of comments a post receives, we'd recommend the post be categorized as ask post and created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est).

However, it should be noted that the data set we analyzed excluded posts without any comments. Given that, it's more accurate to say that of the posts that received comments, ask posts received more comments on average and ask posts created between 15:00 and 16:00 (3:00 pm est - 4:00 pm est) received the most comments on average.